assembly/src/assembler/tests.rs

```diff
@@ -124,90 +125,131 @@ fn nested_blocks() {
+            MastNode::new_split(r#true3, r#false3, expected_mast_forest_builder.forest());
+        expected_mast_forest_builder.ensure_node(r#if_node)
     };
 
     let r#while = {
         let push_basic_block_id = {
             let push_basic_block = MastNode::new_basic_block(vec![Operation::Push(23u32.into())]);
-            expected_mast_forest.ensure_node(push_basic_block)
+            expected_mast_forest_builder.ensure_node(push_basic_block)
         };
         let body_node_id = {
-            let body_node =
-                MastNode::new_join(exec_bar_node_id, push_basic_block_id, &expected_mast_forest);
+            let body_node = MastNode::new_join(
+                exec_bar_node_id,
+                push_basic_block_id,
+                expected_mast_forest_builder.forest(),
+            );
 
-            expected_mast_forest.ensure_node(body_node)
+            expected_mast_forest_builder.ensure_node(body_node)
         };
 
-        let loop_node = MastNode::new_loop(body_node_id, &expected_mast_forest);
-        expected_mast_forest.ensure_node(loop_node)
+        let loop_node = MastNode::new_loop(body_node_id, expected_mast_forest_builder.forest());
+        expected_mast_forest_builder.ensure_node(loop_node)
     };
     let push_13_basic_block_id = {
         let node = MastNode::new_basic_block(vec![Operation::Push(13u32.into())]);
-        expected_mast_forest.ensure_node(node)
+        expected_mast_forest_builder.ensure_node(node)
     };
 
     let r#false2 = {
-        let node = MastNode::new_join(push_13_basic_block_id, r#while, &expected_mast_forest);
-        expected_mast_forest.ensure_node(node)
+        let node = MastNode::new_join(
+            push_13_basic_block_id,
+            r#while,
+            expected_mast_forest_builder.forest(),
+        );
+        expected_mast_forest_builder.ensure_node(node)
     };
     let nested = {
-        let node = MastNode::new_split(r#true2, r#false2, &expected_mast_forest);
-        expected_mast_forest.ensure_node(node)
+        let node = MastNode::new_split(r#true2, r#false2, expected_mast_forest_builder.forest());
+        expected_mast_forest_builder.ensure_node(node)
     };
 
     let combined_node_id = combine_mast_node_ids(
         vec![before, r#if1, nested, exec_foo_bar_baz_node_id, syscall_foo_node_id],
-        &mut expected_mast_forest,
+        &mut expected_mast_forest_builder,
     );
-    expected_mast_forest.set_entrypoint(combined_node_id);
 
-    let combined_node = &expected_mast_forest[combined_node_id];
+    let expected_program = Program::new(expected_mast_forest_builder.build(), combined_node_id);
+    assert_eq!(expected_program.hash(), program.hash());
+
+    // also check that the program has the right number of procedures
+    assert_eq!(program.num_procedures(), 5);
+}
+
+/// Ensures that a single copy of procedures with the same MAST root are added only once to the MAST
+/// forest.
+#[test]
+fn duplicate_procedure() {
```

Comment:
I would also add a test to check that duplicate nodes get collapsed into a single node. For example:
```
begin
    push.3.2.1.0
    if.true
        mul
    else
        if.true add else mul end
    end
end
```
In the above the the `if.true mul else` and `else mul end` should refer to the same node as in both cases the basic block is just a single `mul` instruction.

Reply:
Done in [9d48fda](https://github.com/0xPolygonMiden/miden-vm/pull/1359/commits/9d48fdaef52523861e10630f24b6e2ddd5233ab6)